plan/logical_plans.go

```diff
@@ -299,13 +304,86 @@ type DataSource struct {
 
 	statisticTable *statistics.Table
 
-	// availableIndices is used for storing result of availableIndices function.
-	availableIndices *availableIndices
+	// possibleAccessPaths stores all the possible access path for physical plan, including table scan.
+	// Please make sure table path is always the first element if we have table path.
+	possibleAccessPaths []*accessPath
+}
+
+// accessPath tells how we access one index or just access table.
+type accessPath struct {
+	index  *model.IndexInfo
+	ranges []*ranger.Range
+	// countAfterAccess is the row count after we apply range seek and before we use other filter to filter data.
+	countAfterAccess float64
+	// countAfterIndex is the row count after we apply filters on index and before we apply the table filters.
+	countAfterIndex float64
+	accessConds     []expression.Expression
+	eqCondCount     int
+	indexFilters    []expression.Expression
+	tableFilters    []expression.Expression
+	// isTablePath indicates whether this path is table path.
+	isTablePath bool
+	// forced means this path is generated by `use/force index()`.
+	forced bool
+}
+
+func (ds *DataSource) deriveTablePathStats(path *accessPath) error {
+	var err error
+	sc := ds.ctx.GetSessionVars().StmtCtx
+	path.countAfterAccess = float64(ds.statisticTable.Count)
+	path.tableFilters = ds.pushedDownConds
+	var pkCol *expression.Column
+	if ds.tableInfo.PKIsHandle {
+		if pkColInfo := ds.tableInfo.GetPkColInfo(); pkColInfo != nil {
+			pkCol = expression.ColInfo2Col(ds.schema.Columns, pkColInfo)
+		}
+	}
+	if pkCol == nil {
```

Comment:
We'better add some test cases to cover pkCol == nil and pkCol !=nil.

Reply:
the status of pkCol is covered executor's unit test

Reply:
But why all test cases can be passed when line 336~340 was missed before,
we may need some test for checking this case.

Reply:
let me check it first.